### PR TITLE
Ad user story 45 48

### PR DESCRIPTION
**Pull request by adumortier**
- added the test that a `merchant` can `add/edit` an `item` through a form
- added the test that an incomplete `edit/new` form submission takes the `merchant` back to the form
- added the test that the `merchant` gets redirected to pre-populated `edit/new` form
- added the functionality, views and routes for these tests.

### CODE REVIEW

File: app/controllers/merchant/items_controller.rb [hunk 1]
```diff
@@ -7,45 +9,95 @@ def index
...
+    item = Item.find(params[:id])
+    destroy_item(item)
+  end
+
+  def new
+    @item = if session[:failed_save]
+              Item.new(session[:failed_save])
+            else
+              Item.new
+            end
```
Comment (TravisBorgsmiller) [inline app/controllers/merchant/items_controller.rb:40]: Love this session logic

File: app/controllers/merchant/items_controller.rb [hunk 2]
```diff
@@ -7,45 +9,95 @@ def index
...
+  end
+
   def create
-    merchant = Merchant.where("id = #{current_user.merchant.id}").first
+    merchant = Merchant.find(current_user.merchant.id)
     @item = merchant.items.create(item_params)
```
Comment (TravisBorgsmiller) [inline app/controllers/merchant/items_controller.rb:45]: This seems like a cleaner way of grabbing the merchant

File: app/controllers/merchant/items_controller.rb [hunk 3]
```diff
@@ -7,45 +9,95 @@ def index
...
+  end
+
+  def update_item(item)
+    @item = item
+    @item.update(item_params)
     if @item.save
-      flash[:notice] = 'Your new item was saved'
-      redirect_to "/merchant/items"
+      flash[:notice] = 'Your item was updated successfully'
+      redirect_to '/merchant/items'
```
Comment (TravisBorgsmiller) [inline app/controllers/merchant/items_controller.rb:60]: much sexier flash message

**Review by TravisBorgsmiller — APPROVED** [review]

File: spec/features/users/merchant_add_item_spec.rb [hunk 4]
```diff
@@ -122,11 +123,11 @@
...
-        fill_in :_merchant_items_name, with: name
-        fill_in :_merchant_items_price, with: price
-        fill_in :_merchant_items_description, with: description
-        fill_in :_merchant_items_image, with: image_url
-        fill_in :_merchant_items_inventory, with: inventory
+        fill_in :item_name, with: name
+        fill_in :item_price, with: price
+        fill_in :item_description, with: description
+        fill_in :item_image, with: image_url
+        fill_in :item_inventory, with: inventory
```
Comment (cgaddis36) [inline spec/features/users/merchant_add_item_spec.rb:131]: better naming convention than what was previously here, good job on that

File: spec/features/users/merchant_edit_item_spec.rb [hunk 5]
```diff
@@ -0,0 +1,171 @@
...
+    before(:each) do
+        @merchant_user = User.create!(name: 'Alejandro',
+                                      street_address: '321 Jones Dr',
+                                      city: 'Jonesville',
+                                      state: 'WI',
+                                      zip_code: 54_321,
+                                      email: 'alejandro@example.com',
+                                      password: 'hamburger3',
+                                      role: 1)
+        allow_any_instance_of(ApplicationController).to receive(:current_user).and_return(@merchant_user)
```
Comment (cgaddis36) [inline spec/features/users/merchant_edit_item_spec.rb]: I might move this below where you add the merchant_user to @merchant.users just to complete the merchant_user creation functionality before logging in with them

File: spec/features/users/merchant_add_item_spec.rb [hunk 6]
```diff
@@ -122,11 +123,11 @@
...
```
Comment (cgaddis36) [inline spec/features/users/merchant_add_item_spec.rb:130]: nice test for the image_url placeholder

File: spec/features/users/merchant_edit_item_spec.rb [hunk 7]
```diff
@@ -0,0 +1,171 @@
...
+      original_status = @item3.active?
+    
+      expect(current_path).to eq("/merchant/items/#{@item3.id}/edit")
+
+      expect(find_field('Name').value).to eq @item3.name
+      expect(find_field('Price').value).to eq @item3.price.to_s
+      expect(find_field('Description').value).to eq @item3.description
+      expect(find_field('Image').value).to eq @item3.image
+      expect(find_field('Inventory').value).to eq @item3.inventory.to_s
+
```
Comment (cgaddis36) [inline spec/features/users/merchant_edit_item_spec.rb:46]: very explicit field testing... we may want to make sure we test all edit fields this thoroughly in our refactor

File: spec/features/users/merchant_edit_item_spec.rb [hunk 8]
```diff
@@ -0,0 +1,171 @@
...
+      fill_in :item_price, with: price
+      fill_in :item_description, with: description
+      fill_in :item_image, with: image_url
+      fill_in :item_inventory, with: inventory
+
+      click_button "Update Item"
+
+      expect(current_path).to eq("/merchant/items/#{@item3.id}/edit")
+
+      expect(find_field('Name').value).to eq @item3.name
```
Comment (cgaddis36) [inline spec/features/users/merchant_edit_item_spec.rb:161]: I do not see the flash message being testes anywhere here... I would also like to see the edge cases tested for where a price cannot be less than 0 as well as the inventory

File: app/controllers/admin/merchants/items_controller.rb [hunk 9]
```diff
@@ -16,6 +16,7 @@ def update
   end
 
   def destroy
+    require 'pry'; binding.pry
```
Comment (cgaddis36) [inline app/controllers/admin/merchants/items_controller.rb:19]: Pry

File: spec/features/users/merchant_edit_item_spec.rb [hunk 10]
```diff
@@ -0,0 +1,174 @@
...
+      expect(find_field('Price').value).to eq @item3.price.to_s
+      expect(find_field('Description').value).to eq @item3.description
+      expect(find_field('Image').value).to eq @item3.image
+      expect(find_field('Inventory').value).to eq @item3.inventory.to_s  
+
+      expect(page).to have_content("Price can't be blank")
+      expect(page).to have_content("Price can't be blank and Price must be greater than 0")
+
```
Comment (cgaddis36) [inline spec/features/users/merchant_edit_item_spec.rb]: Nice additions looks good!

File: spec/features/users/merchant_edit_item_spec.rb [hunk 11]
```diff
@@ -0,0 +1,174 @@
...
+      click_button "Update Item"
+
+      expect(current_path).to eq("/merchant/items")
+
+      within("section#item-#{@item3.id}") do
+        expect(page).to have_content(name)
+        expect(page).to have_content("Price: $#{price}")
+        expect(page).to have_css("img[src*='#{"https://t3.ftcdn.net/jpg/02/48/42/64/240_F_248426448_NVKLywWqArG2ADUxDq6QprtIzsF82dMF.jpg"}']")
+        expect(page).to have_content("Active")
+        expect(page).to_not have_content(description)
```
Comment (cgaddis36) [inline spec/features/users/merchant_edit_item_spec.rb:117]: good job . testing the default image

File: app/controllers/merchant/items_controller.rb [hunk 12]
```diff
@@ -7,45 +9,95 @@ def index
...
+              Item.find(session[:failed_update])
+            else
+              Item.find(params[:id])
+            end
+  end
+
+  def update
+    @item = Item.find(params[:id])
+    if params[:item]
+      update_item(@item)
```
Comment (cgaddis36) [inline app/controllers/merchant/items_controller.rb:23]: Nice job merging these together

File: app/controllers/merchant/items_controller.rb [hunk 13]
```diff
@@ -7,45 +9,95 @@ def index
...
+    @item = item
+    if @item.save
+      flash[:notice] = 'Your new item was saved'
+      redirect_to '/merchant/items'
     else
-      flash[:success] = "#{item.name} is deactivated"
+      session[:failed_save] = item_params
+      flash[:error] = @item.errors.full_messages.to_sentence
+      redirect_to '/merchant/items/new'
     end
```
Comment (cgaddis36) [inline app/controllers/merchant/items_controller.rb:93]: I like the use of sessions here good work on that last night

File: app/views/merchant/items/edit.html.erb [hunk 14]
```diff
@@ -0,0 +1,23 @@
+<center>
```
Comment (cgaddis36) [inline app/views/merchant/items/edit.html.erb:1]: form_for :)

File: spec/features/users/merchant_edit_item_spec.rb [hunk 15]
```diff
@@ -0,0 +1,169 @@
...
+      click_button 'Update Item'
+
+      expect(current_path).to eq("/merchant/items/#{@item3.id}/edit")
+
+      expect(page).to_not have_content("Name can't be blank")
+      expect(page).to have_content("Description can't be blank")
+      expect(page).to have_content("Price must be greater than 0")
+      expect(page).to have_content('Inventory must be greater than or equal to 0')
+      expect(page).to_not have_content("Image can't be blank")
+
```
Comment (cgaddis36) [inline spec/features/users/merchant_edit_item_spec.rb:160]: Okay looks good now on inventory validation

File: app/controllers/merchant/items_controller.rb (hunk 2 above)
Comment (KellyIB) [reply app/controllers/merchant/items_controller.rb:45]: Agreed!

**Review by cgaddis36 — APPROVED** [review]
Everything looks ready to merge now, everything passes in RSpec!

**Pull request merged**